include/nighthawk/adaptive_load/nighthawk_service_client.h

```diff
@@ -0,0 +1,38 @@
+#pragma once
+#include "envoy/common/pure.h"
+
+#include "external/envoy/source/common/common/statusor.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "api/client/options.pb.h"
+#include "api/client/service.grpc.pb.h"
+
+namespace Nighthawk {
+
+/**
+ * An interface for interacting with a Nighthawk Service gRPC stub.
+ */
+class NighthawkServiceClient {
+public:
+  virtual ~NighthawkServiceClient() = default;
+
+  /**
+   * Runs a single benchmark using a Nighthawk Service.
+   *
+   * @param nighthawk_service_stub Nighthawk Service gRPC stub.
+   * @param command_line_options Nighthawk Service benchmark request proto generated by the
+   * StepController, without the duration set.
+   * @param duration Duration to insert into the benchmark request.
+   *
+   * @return StatusOr<ExecutionResponse> If we reached the Nighthawk Service, this is the raw
+   * ExecutionResponse proto, containing the benchmark data or possibly an error message from
+   * Nighthawk Service; if we had trouble communicating with the Nighthawk Service, we return an
+   * error status.
+   */
+  virtual absl::StatusOr<nighthawk::client::ExecutionResponse> PerformNighthawkBenchmark(
+      nighthawk::client::NighthawkService::StubInterface* nighthawk_service_stub,
```

Comment:
Not a request for change, but I wanted to hear your rationale: why have this as an argument on the function, rather than an argument to the class constructor?

It's interesting to me that neither this nor the impl version seem to hold any state at all, which leads to a question of why they need to be classes at all.

Reply:
This used to be a bare function because it had no need to track state across calls. But in order to mock out this functionality for testing, with gtest it's much simpler to make a library export an interface rather than plain functions.

Leaving it stateless seems simpler than storing the stub and taking on responsibility for its lifetime. The intent here is definitely not to introduce a new layer into the system architecture, only to wrap a stateless helper function in a class for unit test purposes.

Reply:
Makes sense.  Thanks for explaining.